File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.res = -float("inf")
        def dfs(node):
            if not node:
                return 0
            print(node.val)
            left = dfs(node.left) 
            right = dfs(node.right)
            l = max(left, 0)
            r = max(right, 0)
            self.res = max(self.res,l + node.val, r + node.val, left + right + node.val)
            return max(left + node.val, right + node.val)
        dfs(root)
        return self.res

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        auto dfs = [](auto self, TreeNode* root) {
            if (!root) {
                return make_pair(true, 0);
            }
            auto [balancedLeft, heightLeft] = self(self, root->left);
            auto [balancedRight, heightRight] = self(self, root->right);
            bool balanced{balancedLeft && balancedRight && abs(heightLeft - heightRight) <= 1};
            int height{std::max(heightLeft, heightRight) + 1};
            return make_pair(balanced, height);
        };
        return dfs(dfs, root).first;
    }
};
